File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-4.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        ans = 0
        l= 0 

        for r in range(len(s)):
            count[s[r]] = count.get(s[r], 0) + 1
            maxF = max(count.values())

            if (r - l + 1) - maxF <= k:
                ans = max(ans, (r - l + 1))
            else:
        # k too small shrink window
                count[s[l]] -= 1
                if count[s[l]] == 0:
                        del count[s[l]]
                l += 1
            
        return ans

File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        if (heights.size() == 0) {return 0;}
        int ans = 0, n = heights.size(), l = 0, r = n - 1;

        while (l < r ){ 
            int area = min(heights[l], heights[r]) * (r - l);
            ans = max(ans, area);
            if (heights[l] > heights[r]) {
                r--;
            } 
            else {
                l++;
            }
        }
        return ans;
    }
};
